core/shared/src/main/scala/zio/stm/TMap.scala

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.stm
+
+/**
+ * Transactional map implemented on top of [[TRef]] and [[TArray]]. Resolves
+ * conflicts via chaining.
+ *
+ * Caution: doesn't provide stack-safety guarantees.
+ */
+class TMap[K, V] private (
+  private val tBuckets: TRef[TArray[List[(K, V)]]],
+  private val tCapacity: TRef[Int],
+  private val tSize: TRef[Int]
+) {
+
+  /**
+   * Tests whether or not map contains a key.
+   */
+  final def contains(k: K): STM[Nothing, Boolean] =
+    get(k).map(_.isDefined)
+
+  /**
+   * Removes binding for given key.
+   */
+  final def delete(k: K): STM[Nothing, Unit] = {
+    def removeMatching(bucket: List[(K, V)]): STM[Nothing, List[(K, V)]] = {
+      val (toRemove, toRetain) = bucket.partition(_._1 == k)
+      if (toRemove.isEmpty) STM.succeed(toRetain) else tSize.update(_ - toRemove.size).as(toRetain)
+    }
+
+    for {
+      buckets <- tBuckets.get
+      idx     <- indexOf(k)
+      _       <- buckets.updateM(idx, removeMatching)
+    } yield ()
+  }
+
+  /**
+   * Atomically folds using pure function.
+   */
+  final def fold[A](zero: A)(op: (A, (K, V)) => A): STM[Nothing, A] =
+    for {
+      buckets <- tBuckets.get
+      res     <- buckets.fold(zero)((acc, bucket) => bucket.foldLeft(acc)(op))
+    } yield res
+
+  /**
+   * Atomically folds using effectful function.
+   */
+  final def foldM[A, E](zero: A)(op: (A, (K, V)) => STM[E, A]): STM[E, A] = {
+    def loopM(acc: STM[E, A], remaining: List[(K, V)]): STM[E, A] =
+      remaining match {
+        case Nil          => acc
+        case head :: tail => loopM(acc.flatMap(op(_, head)), tail)
+      }
+
+    for {
+      buckets <- tBuckets.get
+      res     <- buckets.foldM(zero)((acc, bucket) => loopM(STM.succeed(acc), bucket))
+    } yield res
+  }
+
+  /**
+   * Atomically performs side-effect for each binding present in map.
+   */
+  final def foreach[E](f: ((K, V)) => STM[E, Unit]): STM[E, Unit] =
```

Comment:
Nested tuple `((K, V))` instead of 2 param function here will make it harder to use in general.

Reply:
Good catch, removed.